homeassistant/components/unifi/entity.py

```diff
@@ -0,0 +1,176 @@
+"""UniFi entity representation."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Generic, TypeVar
+
+import aiounifi
+from aiounifi.interfaces.api_handlers import CallbackType, ItemEvent, UnsubscribeType
+from aiounifi.interfaces.devices import Devices
+from aiounifi.models.device import Device
+from aiounifi.models.event import Event, EventKey
+
+from homeassistant.core import callback
+from homeassistant.helpers import entity_registry as er
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity import DeviceInfo, Entity, EntityDescription
+
+from .controller import UniFiController
+
+DataT = TypeVar("DataT", bound=Device)
+HandlerT = TypeVar("HandlerT", bound=Devices)
+SubscriptionT = Callable[[CallbackType, ItemEvent], UnsubscribeType]
+
+
+@dataclass
+class UnifiDescription(Generic[HandlerT, DataT]):
+    """Validate and load entities from different UniFi handlers."""
+
+    allowed_fn: Callable[[UniFiController, str], bool]
+    api_handler_fn: Callable[[aiounifi.Controller], HandlerT]
+    available_fn: Callable[[UniFiController, str], bool]
+    device_info_fn: Callable[[aiounifi.Controller, str], DeviceInfo | None]
+    event_is_on: tuple[EventKey, ...] | None
+    event_to_subscribe: tuple[EventKey, ...] | None
+    name_fn: Callable[[DataT], str | None]
+    object_fn: Callable[[aiounifi.Controller, str], DataT]
+    supported_fn: Callable[[UniFiController, str], bool | None]
+    unique_id_fn: Callable[[UniFiController, str], str]
+
+
+@dataclass
+class UnifiEntityDescription(EntityDescription, UnifiDescription[HandlerT, DataT]):
+    """UniFi Entity Description."""
+
+
+class UnifiEntity(Entity, Generic[HandlerT, DataT]):
+    """Representation of a UniFi entity."""
+
+    entity_description: UnifiEntityDescription[HandlerT, DataT]
+    _attr_should_poll = False
+
+    _attr_unique_id: str
+
+    def __init__(
+        self,
+        obj_id: str,
+        controller: UniFiController,
+        description: UnifiEntityDescription[HandlerT, DataT],
+    ) -> None:
+        """Set up UniFi switch entity."""
+        self._obj_id = obj_id
+        self.controller = controller
+        self.entity_description = description
+
+        self._removed = False
+
+        self._attr_available = description.available_fn(controller, obj_id)
+        self._attr_device_info = description.device_info_fn(controller.api, obj_id)
+        self._attr_unique_id = description.unique_id_fn(controller, obj_id)
+
+        obj = description.object_fn(self.controller.api, obj_id)
+        self._attr_name = description.name_fn(obj)
+        self.async_initiate_state()
+
+    async def async_added_to_hass(self) -> None:
```

Comment:
There was previously a subscription like this:
```py
        self.async_on_remove(
            async_dispatcher_connect(
                self.hass,
                self.controller.signal_remove,
                self.remove_item,
            )
        )
```
Is that no longer needed?

Reply:
Exactly, this is replaced by the `if event == ItemEvent.DELETED and obj_id == self._obj_id:`